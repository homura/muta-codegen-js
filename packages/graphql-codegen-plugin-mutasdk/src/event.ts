import { PluginFunction } from '@graphql-codegen/plugin-helpers';

const importSegment = `
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec } from "@mutajs/types";
`;

interface EventDef {
  name: string;
}

function template(config: { events: EventDef[] }): string {
  const events = config.events;

  if (!events.length) return '';

  // prettier-ignore
  return `
${events.map(({ name }) => `
export function isEventOf${name}(event: any): event is ${name} {
  return event?.topic === '${name}';
}`).join('\n')}

type _EventName = ${events.map(({ name }) => `'${name}'`).join(' |\n')};

interface _IsEventOf {
  ${events.map(({ name }) =>
    `(name: '${name}', event: any): event is ${name};`
  ).join('\n')}
}

export const isEventOf =
  ((name: _EventName, event: any) => event?.topic === name) as _IsEventOf;
  `;
}

export const plugin: PluginFunction = schema => {
  const eventDefs = Object.keys(schema.getTypeMap())
    .filter(name => name.endsWith('Event'))
    .map<EventDef>(name => ({ name }));

  return {
    prepend: [importSegment],
    content: template({ events: eventDefs }),
  };
};
