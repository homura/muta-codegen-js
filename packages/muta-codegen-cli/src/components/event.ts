import * as typescriptPlugin from '@graphql-codegen/typescript';
import { event as eventPlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateEventCode(eventSource: string): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      event: eventPlugin,
    },
    plugins: [{ event: {} }, { typescript: {} }],
    schemaSource: eventSource,
  });
}
