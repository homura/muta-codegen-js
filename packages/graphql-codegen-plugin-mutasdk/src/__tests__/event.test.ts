import { buildSchema } from 'graphql';
import { generateEventDefCode } from '../event';

it('empty event input should output empty result', () => {
  expect(generateEventDefCode(buildSchema('scalar Hash'))).toEqual('');
});

it('test one event', () => {
  expect(
    generateEventDefCode(
      buildSchema(`
  scalar Hash
  scalar Address
  scalar u64

  type CreateAssetEvent {
    asset_id: Hash!
    from: Address!
    to: Address!
    value: u64
  }
  `),
    ),
  ).toMatchSnapshot();
});
