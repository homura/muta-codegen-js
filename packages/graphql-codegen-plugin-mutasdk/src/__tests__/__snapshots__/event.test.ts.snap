// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test one event 1`] = `
"

export function isEventOfCreateAssetEvent(event: any): event is CreateAssetEvent {
  return event?.topic === 'CreateAssetEvent';
}

type _EventName = 'CreateAssetEvent';

interface _IsEventOf {
  (name: 'CreateAssetEvent', event: any): event is CreateAssetEvent;
}

export const isEventOf: _IsEventOf =
  ((name: _EventName, event: any) => event?.topic === name) as _IsEventOf;
  "
`;
